platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,470 @@
+
+        final long start = System.currentTimeMillis();
+
+        final long filesToCompactSize = getSizeOfFiles(filesToCompact);
+        logger.debug(
+                MERKLE_DB.getMarker(),
+                "[{}] Starting merging {} files / {}",
+                storeName,
+                filesCount,
+                formatSizeBytes(filesToCompactSize));
+
+        final List<Path> newFilesCreated = compactFiles(index, filesToCompact, targetCompactionLevel);
+
+        final long end = System.currentTimeMillis();
+        final long tookMillis = end - start;
+        if (reportDurationMetricFunction != null) {
+            reportDurationMetricFunction.accept(targetCompactionLevel, tookMillis);
+        }
+
+        final long compactedFilesSize = getSizeOfFilesByPath(newFilesCreated);
+        if (reportSavedSpaceMetricFunction != null) {
+            reportSavedSpaceMetricFunction.accept(
+                    targetCompactionLevel,
+                    (filesToCompactSize - compactedFilesSize) * UnitConstants.BYTES_TO_MEBIBYTES);
+        }
+
+        logCompactStats(
+                storeName,
+                tookMillis,
+                filesToCompact,
+                filesToCompactSize,
+                newFilesCreated,
+                targetCompactionLevel,
+                dataFileCollection);
+        logger.warn(
+                MERKLE_DB.getMarker(),
+                "[{}] Finished compaction {} files / {} in {} ms",
+                storeName,
+                filesCount,
+                formatSizeBytes(filesToCompactSize),
+                tookMillis);
+
+        return true;
+    }
+
+    /**
+     * This method creates a compaction plan (a set of files to be compacted). The plan is organized by compaction levels
+     * in ascending order. If there are not enough files to compact, then no files are compacted and the plan will be empty.
+     * If the current level doesn't reach minNumberOfFilesToCompact threshold,
+     * then this level and the levels above it are not included in the plan.
+     * @return filter creating a compaction plan
+     */
+    List<DataFileReader<?>> compactionPlan(List<DataFileReader<?>> dataFileReaders) {
+        if (dataFileReaders.isEmpty()) {
+            return dataFileReaders;
+        }
+        final Map<Integer, List<DataFileReader<?>>> readersByLevel = new HashMap<>();
+        int maxCompactionLevel = config.maxCompactionLevel();
+        for (int i = 0; i <= maxCompactionLevel; i++) {
+            readersByLevel.put(i, new ArrayList<>());
+        }
+
+        for (DataFileReader<?> reader : dataFileReaders) {
+            readersByLevel.get(reader.getMetadata().getCompactionLevel()).add(reader);
+        }
+
+        final List<DataFileReader<?>> readersToCompact = new ArrayList<>();
+
+        List<DataFileReader<?>> nonCompactedReaders = readersByLevel.get(DEFAULT_COMPACTION_LEVEL);
```

Comment:
```java
for (int i = 0; i <= maxCompactionLevel; i++) {
    final List<DataFileReader<?>> readers = readersByLevel.get(i);
    final int minRequiredAtLevel = (i == 0) ? getMinNumberOfFilesToCompact() : getMinNumberOfFilesToCompact() - 1;
    if (readers.size() < minRequiredAtLevel) {
        break;
    }
    readersToCompact.addAll(readers);
}
```

I'm not sure if it's more readable or not, leaving it up to you

Reply:
I updated it